# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/cookiestore/CookieStore.cpp
# Commit: 675785853d7a
# Full Hash: 675785853d7a40123826e1b47c49d68f87bd4bd0
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2024-12-17 17:00:49
# Regressor Bug: 1915355
# File Overlap Count: 1
# Description:
#   Bug 1915355 - Part 5: Exempt third-party cookie blocking if the channel is on the exception list. r=cookie-reviewers,bvandersloot,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223384
# ==============================================================================

diff -r de6a26d3e3df -r 675785853d7a dom/cookiestore/CookieStore.cpp
--- a/dom/cookiestore/CookieStore.cpp	Tue Dec 17 08:57:31 2024 +0000
+++ b/dom/cookiestore/CookieStore.cpp	Tue Dec 17 08:57:31 2024 +0000
@@ -196,11 +196,12 @@
 
 bool GetContextAttributes(CookieStore* aCookieStore, bool* aThirdPartyContext,
                           bool* aPartitionForeign, bool* aUsingStorageAccess,
-                          Promise* aPromise) {
+                          bool* aIsOn3PCBExceptionList, Promise* aPromise) {
   MOZ_ASSERT(aCookieStore);
   MOZ_ASSERT(aThirdPartyContext);
   MOZ_ASSERT(aPartitionForeign);
   MOZ_ASSERT(aUsingStorageAccess);
+  MOZ_ASSERT(aIsOn3PCBExceptionList);
   MOZ_ASSERT(aPromise);
 
   if (NS_IsMainThread()) {
@@ -224,6 +225,7 @@
 
     *aPartitionForeign = document->CookieJarSettings()->GetPartitionForeign();
     *aUsingStorageAccess = document->UsingStorageAccess();
+    *aIsOn3PCBExceptionList = document->IsOn3PCBExceptionList();
     return true;
   }
 
@@ -234,6 +236,7 @@
   *aPartitionForeign =
       workerPrivate->CookieJarSettings()->GetPartitionForeign();
   *aUsingStorageAccess = workerPrivate->UsingStorageAccess();
+  *aIsOn3PCBExceptionList = workerPrivate->IsOn3PCBExceptionList();
   return true;
 }
 
@@ -376,9 +379,11 @@
         bool thirdPartyContext = true;
         bool partitionForeign = true;
         bool usingStorageAccess = false;
+        bool isOn3PCBExceptionList = false;
 
         if (!GetContextAttributes(self, &thirdPartyContext, &partitionForeign,
-                                  &usingStorageAccess, promise)) {
+                                  &usingStorageAccess, &isOn3PCBExceptionList,
+                                  promise)) {
           return;
         }
 
@@ -407,8 +412,8 @@
                 aOptions.mDomain.IsEmpty() ? nsString(baseDomain)
                                            : nsString(aOptions.mDomain),
                 cookiePrincipal->OriginAttributesRef(), thirdPartyContext,
-                partitionForeign, usingStorageAccess, nsString(aOptions.mName),
-                nsString(aOptions.mValue),
+                partitionForeign, usingStorageAccess, isOn3PCBExceptionList,
+                nsString(aOptions.mName), nsString(aOptions.mValue),
                 // If expires is not set, it's a session cookie.
                 aOptions.mExpires.IsNull(),
                 aOptions.mExpires.IsNull()
@@ -499,9 +504,11 @@
         bool thirdPartyContext = true;
         bool partitionForeign = true;
         bool usingStorageAccess = false;
+        bool isOn3PCBExceptionList = false;
 
         if (!GetContextAttributes(self, &thirdPartyContext, &partitionForeign,
-                                  &usingStorageAccess, promise)) {
+                                  &usingStorageAccess, &isOn3PCBExceptionList,
+                                  promise)) {
           return;
         }
 
@@ -530,8 +537,9 @@
                 aOptions.mDomain.IsEmpty() ? nsString(baseDomain)
                                            : nsString(aOptions.mDomain),
                 cookiePrincipal->OriginAttributesRef(), thirdPartyContext,
-                partitionForeign, usingStorageAccess, nsString(aOptions.mName),
-                path, aOptions.mPartitioned, operationID);
+                partitionForeign, usingStorageAccess, isOn3PCBExceptionList,
+                nsString(aOptions.mName), path, aOptions.mPartitioned,
+                operationID);
         if (NS_WARN_IF(!ipcPromise)) {
           promise->MaybeResolveWithUndefined();
           return;
@@ -705,9 +713,11 @@
         bool thirdPartyContext = true;
         bool partitionForeign = true;
         bool usingStorageAccess = false;
+        bool isOn3PCBExceptionList = false;
 
         if (!GetContextAttributes(self, &thirdPartyContext, &partitionForeign,
-                                  &usingStorageAccess, promise)) {
+                                  &usingStorageAccess, &isOn3PCBExceptionList,
+                                  promise)) {
           return;
         }
 
@@ -731,8 +741,8 @@
                     ? Some(partitionedCookiePrincipal->OriginAttributesRef())
                     : Nothing(),
                 thirdPartyContext, partitionForeign, usingStorageAccess,
-                aOptions.mName.WasPassed(), nsString(name), path,
-                aOnlyTheFirstMatch);
+                isOn3PCBExceptionList, aOptions.mName.WasPassed(),
+                nsString(name), path, aOnlyTheFirstMatch);
         if (NS_WARN_IF(!ipcPromise)) {
           promise->MaybeResolveWithUndefined();
           return;